File: internal/chezmoi/init_test.go
package chezmoi

import (
	"strings"
	"testing"
)

func TestRepoURL(t *testing.T) {
	if got := RepoURL("octocat"); got != "git@github.com:octocat/dotfiles.git" {
		t.Errorf("RepoURL(octocat) = %q", got)
	}
	if got := RepoURL(""); !strings.Contains(got, "<you>") {
		t.Errorf("empty user should leave a placeholder, got %q", got)
	}
}

func TestPlanInitAllDone(t *testing.T) {
	steps := PlanInit(InitState{ChezmoiInstalled: true, AgeKeyConfigured: true, SourceInitialized: true})
	if !IsReady(steps) {
		t.Fatal("all-satisfied state should be ready")
	}
	for _, s := range steps {
		if s.Command != "" {
			t.Errorf("done step %q should carry no command, got %q", s.ID, s.Command)
		}
	}
}

func TestPlanInitNoneDone(t *testing.T) {
	steps := PlanInit(InitState{User: "octocat"})
	if IsReady(steps) {
		t.Fatal("empty state should not be ready")
	}
	cmds := map[string]string{}
	for _, s := range steps {
		cmds[s.ID] = s.Command
	}
	if cmds["chezmoi"] != "brew install chezmoi" {
		t.Errorf("chezmoi step command = %q", cmds["chezmoi"])
	}
	if !strings.Contains(cmds["age-key"], "age-keygen") {
		t.Errorf("age-key step command = %q", cmds["age-key"])
	}
	if cmds["source"] != "chezmoi init git@github.com:octocat/dotfiles.git" {
		t.Errorf("source step command = %q", cmds["source"])
	}
	// The key-loss warning must be present.
	for _, s := range steps {
		if s.ID == "age-key" && !strings.Contains(s.Note, "unrecoverable") {
			t.Errorf("age-key step missing loss warning: %q", s.Note)
		}
	}
}

File: tests/commands/init.test.ts
import { test, expect, describe } from "bun:test";
import { planInit, isReady, repoUrl, type InitState } from "../../src/commands/init";

const base: InitState = {
  chezmoiInstalled: false,
  ageKeyConfigured: false,
  sourceInitialized: false,
};

describe("repoUrl", () => {
  test("uses the gh login; private repo is named 'dotfiles' (chezmoi convention)", () => {
    expect(repoUrl("doguyilmaz")).toBe("git@github.com:doguyilmaz/dotfiles.git");
  });
  test("falls back to a placeholder when the user is unknown", () => {
    expect(repoUrl()).toBe("git@github.com:<you>/dotfiles.git");
  });
});

describe("planInit", () => {
  test("nothing set up → all three steps todo with their commands", () => {
    const steps = planInit(base);
    expect(steps.map((s) => s.id)).toEqual(["chezmoi", "age-key", "source"]);
    expect(steps.every((s) => s.status === "todo")).toBe(true);
    expect(steps[0].command).toBe("brew install chezmoi");
    expect(steps[1].command).toContain("age-keygen");
    expect(steps[1].note).toContain("Back this key up"); // loss warning present
    expect(steps[2].command).toContain("chezmoi init");
  });

  test("uses the resolved user in the chezmoi init command", () => {
    const steps = planInit({ ...base, chezmoiInstalled: true, ageKeyConfigured: true, user: "doguyilmaz" });
    const source = steps.find((s) => s.id === "source");
    expect(source?.command).toBe("chezmoi init git@github.com:doguyilmaz/dotfiles.git");
  });

  test("done steps carry no command or note", () => {
    const steps = planInit({ chezmoiInstalled: true, ageKeyConfigured: true, sourceInitialized: true });
    expect(steps.every((s) => s.status === "done")).toBe(true);
    expect(steps.every((s) => s.command === undefined && s.note === undefined)).toBe(true);
  });
});

describe("isReady", () => {
  test("true only when every prerequisite is done", () => {
    expect(isReady(planInit(base))).toBe(false);
    expect(isReady(planInit({ chezmoiInstalled: true, ageKeyConfigured: true, sourceInitialized: true }))).toBe(true);
    expect(isReady(planInit({ ...base, chezmoiInstalled: true }))).toBe(false);
  });
});
